cmd/osmosisd/cmd/root.go

```diff
@@ -47,36 +56,181 @@ import (
-		ValidArgs:             []string{"bash", "zsh", "fish", "powershell"},
-		Args:                  cobra.ExactArgs(1),
-		Run: func(cmd *cobra.Command, args []string) {
-			switch args[0] {
-			case "bash":
-				_ = cmd.Root().GenBashCompletion(os.Stdout)
-			case "zsh":
-				_ = cmd.Root().GenZshCompletion(os.Stdout)
-			case "fish":
-				_ = cmd.Root().GenFishCompletion(os.Stdout, true)
-			case "powershell":
-				_ = cmd.Root().GenPowerShellCompletionWithDesc(os.Stdout)
+type Trace struct {
+	Type         string `json:"type"`
+	Counterparty struct {
+		BaseDenom string `json:"base_denom"`
+	} `json:"counterparty"`
+}
+
+type DenomUnitMap struct {
+	Base       string
+	DenomUnits []DenomUnit `json:"denom_units"`
+}
+
+var (
+	//go:embed "osmosis-1-assetlist.json" "osmo-test-5-assetlist.json"
+	assetFS   embed.FS
+	mainnetId = "osmosis-1"
+	testnetId = "osmo-test-5"
+)
+
+func loadAssetList(initClientCtx client.Context, cmd *cobra.Command, basedenomToIBC, IBCtoBasedenom bool) (map[string]DenomUnitMap, map[string]string) {
+	var assetList AssetList
+
+	chainId := GetChainId(initClientCtx, cmd)
+
+	fileName := ""
+	if chainId == mainnetId || chainId == "" {
+		fileName = "cmd/osmosisd/cmd/osmosis-1-assetlist-manual.json"
+	} else if chainId == testnetId {
+		fileName = "cmd/osmosisd/cmd/osmo-test-5-assetlist-manual.json"
+	} else {
+		return nil, nil
+	}
+
+	// The order of precedence for asset list is:
+	//  - If the manually generated asset list (generated via the `update-asset-list` cli cmd) exists (noted by -manual.json ending), use it.
+	//  - If the manually generated asset list does not exist, fall back to the embedded asset list.
+	localFile, err := os.Open(fileName)
+	if err != nil {
+		// If we can't open the local file, fall back to the embedded file.
+		if chainId == mainnetId || chainId == "" {
+			fileName = "osmosis-1-assetlist.json"
+		} else if chainId == testnetId {
+			fileName = "osmo-test-5-assetlist.json"
+		} else {
+			return nil, nil
+		}
+		embeddedFile, err := assetFS.Open(fileName)
+		if err != nil {
+			return nil, nil
+		}
+		defer embeddedFile.Close()
+
+		byteValue, _ := io.ReadAll(embeddedFile)
+		err = json.Unmarshal(byteValue, &assetList)
+		if err != nil {
+			return nil, nil
+		}
+	} else {
+		// If the local file opens successfully, use it instead of the embedded file.
+		defer localFile.Close()
+		byteValue, _ := io.ReadAll(localFile)
+		err = json.Unmarshal(byteValue, &assetList)
+		if err != nil {
+			return nil, nil
+		}
+	}
+
+	baseMap := make(map[string]DenomUnitMap)
+	baseMapRev := make(map[string]string)
+
+	if basedenomToIBC {
+		for _, asset := range assetList.Assets {
+			DenomUnitMap := DenomUnitMap{
+				Base:       asset.Base,
+				DenomUnits: asset.DenomUnits,
 			}
-		},
-	})
+			baseMap["base"+strings.ToLower(asset.Symbol)] = DenomUnitMap
+		}
+	}
+	if IBCtoBasedenom {
+		for _, asset := range assetList.Assets {
+			baseMapRev[asset.Base] = "base" + strings.ToLower(asset.Symbol)
```

Comment:
lol this is def silly, in a followup we'll figure out how to either get this properly or amend chain registry to make it doable